### PR TITLE
[inductor] Add split scan kernel

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119430
* __->__ #117992
* #117991

This PR adds a new type of triton kernel in which data is persistent but the
reduction dimension is split over multiple blocks (up to the entire kernel).
though this is called a reduction dimension, in actuality we only support scans.
because of this limitation, i have to be able to block fusions of split scan
operations with reductions so chose to add a new `ir.SplitScan` node which
is identical but allows for differentiation in the scheduler.

The split scan kernel is also the first to require an additional workspace buffer
which is used to communicate between cuda blocks. this is slightly tricky as we
the exact scratch space requirement isn't known until the grid size is calculated.
here i workaround the issue by setting a minimum rblock size and always allocating
to the maximum possible grid size for a given input tensor.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by peterbell10** [comment]
Here is an example of the code generated for:
```python
@torch.compile
def fn(x):
    return x.cumsum(1)

a = torch.randn(10, 128*1024, device="cuda")
fn(a)
```

<details><summary>generated  code</summary>
<p>

```
# kernel path: /tmp/torchinductor_pbell/zg/czgvmtvrqsiol2vowwfgrys625y3tq3tggc6csnqjjycr6x45sfg.py
# Source Nodes: [cumsum], Original ATen: [aten.cumsum]
# cumsum => cumsum
triton_spl_fused_cumsum_0 = async_compile.triton('triton_', '''
import triton
import triton.language as tl
from torch._inductor.ir import ReductionHint
from torch._inductor.ir import TileHint
from torch._inductor.triton_heuristics import AutotuneHint, split_scan
from torch._inductor.utils import instance_descriptor
from torch._inductor import triton_helpers

@triton.jit
def _triton_helper_fn0(arg0, arg1):
    tmp0 = arg0 + arg1
    return tmp0

@split_scan(
    size_hints=[16, 131072],
    reduction_hint=ReductionHint.INNER,
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*fp32', 2: '*i8', 3: 'i32', 4: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 4), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=(2, 4))]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_spl_fused_cumsum_0', 'mutated_arg_names': [], 'no_x_dim': True}
)
@triton.jit
def triton_(in_ptr0, out_ptr0, ws_ptr, xnumel, rnumel, RBLOCK : tl.constexpr):
    xnumel = 10
    XBLOCK: tl.constexpr = 1
    rnumel = 131072
    xoffset = tl.program_id(1) * XBLOCK
    xindex = tl.full([1], xoffset, tl.int32)
    xmask = xindex < xnumel
    roffset = tl.program_id(0) * RBLOCK
    rindex = roffset + tl.arange(0, RBLOCK)[:]
    rmask = rindex < rnumel
    r1 = rindex
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (r1 + (131072*x0)), xmask, eviction_policy='evict_last', other=0.0)
    tmp1 = tl.num_programs(0)
    tmp2 = ws_ptr.to(tl.pointer_type(tl.uint64)) + xoffset * 1 * tmp1
    tmp3 = tmp0.to(tl.float32)
    tmp4 = tl.broadcast_to(tmp3, [RBLOCK])
    tmp5 = tl.full([], 0, tl.float32)
    tmp6 = tl.where(xmask, tmp4, tmp5)
    tmp7 = tl.reduce(tmp6, 0, _triton_helper_fn0)
    tmp8 = triton_helpers.exclusive_scan_decoupled_lookback(
        tmp2,
        tmp7,
        tl.program_id(0),
        _triton_helper_fn0,
        tmp5,
        DTYPE_VALUE=tl.float32,
        DTYPE_VALUE_AS_UINT=tl.uint32,
        DTYPE_PACK=tl.uint64,
    )
    tmp9 = tl.associative_scan(tmp6, 0, _triton_helper_fn0)
    tmp10 = _triton_helper_fn0(tmp8, tmp9)
    tl.store(out_ptr0 + (r1 + (131072*x0)), tmp10, xmask)
''')

import triton
import triton.language as tl
from torch._inductor.triton_heuristics import grid, split_scan_grid, start_graph, end_graph
from torch._C import _cuda_getCurrentRawStream as get_raw_stream


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (10, 131072), (131072, 1))
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        buf0 = empty((10, 131072), device='cuda', dtype=torch.float32)
        # Source Nodes: [cumsum], Original ATen: [aten.cumsum]
        workspace = empty((40960, ), device='cuda', dtype=torch.uint8)
        workspace.zero_()
        stream0 = get_raw_stream(0)
        triton_spl_fused_cumsum_0.run(arg0_1, buf0, workspace, 10, 131072, grid=split_scan_grid(10, 131072), stream=stream0)
        del arg0_1
        return (buf0, )
```

</p>
</details> 

As it turns out, this runs 5x faster than eager because eager doesn't use cub for batched scans. For single dimension scans we are in the same ballpark as `cub` but often slightly slower. I also tried `rand(128*1024, 10, device="cuda").cumsum(0)` and this is 250x faster than eager (lol).

Note I'm referring to timings reported by `nvprof`, so ignoring the overheads added by pt2.

**Review by lezcano — COMMENTED** [review]
The general structure of the PR looks good. I'm tagging @jansel as the main reviewer given that this is a big one.

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -1176,6 +1215,86 @@ def fn(a):
...
+            inp = _large_cumprod_input(
+                (10, 10000), dim=1, dtype=dtype, device=self.device
+            )
+            self.common(
+                fn,
+                (inp,),
+                reference_in_float=True,
+            )
+
+    @skipCUDAIf(TEST_WITH_ROCM, "Numerical issues on ROCm")
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply test/inductor/test_torchinductor.py:1285]: Very bad:
```
2024-02-01T13:53:01.8550798Z Mismatched elements: 3702448 / 3717120 (99.6%)
2024-02-01T13:53:01.8551339Z Greatest absolute difference: inf at index (14672,) (up to 1e-05 allowed)
2024-02-01T13:53:01.8551928Z Greatest relative difference: inf at index (14672,) (up to 0.002 allowed)
```

Note that all scans are falling back to eager on ROCm though, so there must be a problem with the eager implementation.

[the remark this answers is not in the document]
Comment (lezcano) [reply test/inductor/test_torchinductor.py:1277]: cc @jeffdaily

[the remark this answers is not in the document]
Comment (jeffdaily) [reply test/inductor/test_torchinductor.py:1277]: Are there existing unit tests for the eager implementation that are also failing this drastically on ROCm?

[the remark this answers is not in the document]
Comment (jeffdaily) [reply test/inductor/test_torchinductor.py:1277]: I have forwarded this PR on to our inductor/PT2 team.

**Comment by jansel** [comment]
Looks like some ROCm tests are still failing.

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -1106,6 +1106,15 @@ def generate_numel_expr(self, kernel_name: str, tree):
         # it suffices as a type hint for the purposes of producing the correct code for this type.
         return SymbolicCallArg(expr, tree.numel)
 
+    def generate_workspace_allocation(self, nbytes, device, zero_fill):
+        function = "zeros" if zero_fill else "empty"
```
Comment (jansel) [inline torch/_inductor/codegen/wrapper.py:1110]: unused?

File: torch/_inductor/codegen/triton_split_scan.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        scratch_elems_per_block = 3 if element_nbits == 64 else 1
+        scratch_nbytes_per_block = scratch_elems_per_block * (
+            scratch_type_triton.primitive_bitwidth // 8
+        )
+
+        cse_load = functools.partial(self.cse.generate, self.loads)
+        cse_compute = functools.partial(self.cse.generate, self.compute)
+
+        assert len(self.numels) == 2, "Unexpected tiling"
+        # TODO: Enforce this in split_scan heuristic
```
Comment (jansel) [inline torch/_inductor/codegen/triton_split_scan.py:88]: Isn't this already enforced below?

**Review by jansel — APPROVED** [review]

File: test/inductor/test_torchinductor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jataylo) [reply test/inductor/test_torchinductor.py:1287]: Revisiting the initial scan PR https://github.com/pytorch/pytorch/pull/106581 to debug this further but I can confirm we are falling back to eager mode for scan.

If I disable the fallback option in lowering we see
```
torch._dynamo.exc.BackendCompilerFailed: backend='compile_fx_wrapper' raised:
LoweringException: AssertionError: Found <class 'NoneType'>, which is not a supported top level IR node. See [Note: Inductor IR]
  target: aten.cumsum.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg0_1', layout=FixedLayout('cuda', torch.float32, size=[16, 32], stride=[32, 1]))
  ))
  args[1]: 0
```

[the remark this answers is not in the document]
Comment (jataylo) [reply test/inductor/test_torchinductor.py:1287]: Update: Hoping to enable scan lowering on ROCm via https://github.com/pytorch/pytorch/pull/119369

But for now we still have issues with the split scan kernel's due to some issue with ROCm's triton
```
  File "/root/pytorch/torch/_inductor/codecache.py", line 2375, in _load_kernel
    kernel.precompile()
  File "/root/pytorch/torch/_inductor/triton_heuristics.py", line 204, in precompile
    compiled_binary, launcher = self._precompile_config(
  File "/root/pytorch/torch/_inductor/triton_heuristics.py", line 358, in _precompile_config
    binary = triton.compile(*compile_args, **compile_kwargs)
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/triton/compiler/compiler.py", line 552, in compile
    next_module = compile_kernel(module)
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/triton/compiler/compiler.py", line 427, in <lambda>
    lambda src: optimize_ttir(ast_to_ttir(src, signature, configs[0], constants, debug=debug, arch=arch), arch))
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/triton/compiler/code_generator.py", line 1151, in ast_to_ttir
    raise CompilationError(fn.src, node, repr(e)) from e
torch._dynamo.exc.BackendCompilerFailed: backend='compile_fx_wrapper' raised:
CompilationError: at 17:26:    xindex = tl.full([1], xoffset, tl.int32)
    xmask = xindex < xnumel
    roffset = tl.program_id(0) * RBLOCK
    rindex = roffset + tl.arange(0, RBLOCK)[:]
    rmask = rindex < rnumel
    r0 = rindex
    tmp0 = tl.load(in_ptr0 + (r0), rmask, eviction_policy='evict_last', other=0.0).to(tl.float32)
    tmp1 = tl.num_programs(0)
    tmp2 = ws_ptr.to(tl.pointer_type(tl.uint64)) + xoffset * 1 * tmp1
    tmp3 = tmp0.to(tl.float32)
    tmp4 = tl.broadcast_to(tmp3, [RBLOCK])
    tmp5 = tl.full([], 0, tl.float32)
                          ^
TypeError('0d block_type is forbidden')
```

I will work with our triton team on this issue, for now best to skip us on split scan kernel and we will look at enabling asap.

**Pull request closed without merging**